File: docx-wasm/js/section-property.ts
import { DocGridType } from ".";
import { Footer } from "./footer";
import { Header } from "./header";

export type DocGrid = {
  gridType: DocGridType;
  linePitch?: number;
  charSpace?: number;
};

export class SectionProperty {
  _pageSize: PageSize = {
    w: 11906,
    h: 16838,
  };
  _pageMargin: PageMargin | null = null;
  _docGrid: DocGrid | null = null;
  _header: Header | null = null;
  _firstHeader: Header | null = null;
  _evenHeader: Header | null = null;
  _footer: Footer | null = null;
  _firstFooter: Footer | null = null;
  _evenFooter: Footer | null = null;

  pageSize(w: number, h: number) {
    this._pageSize.w = w;
    this._pageSize.h = h;
    return this;
  }

  pageMargin(margin: Partial<PageMargin>) {
    this._pageMargin = { ...defaultPageMargin(), ...margin };
    return this;
  }

  pageOrientation(orient: PageOrientationType) {
    this._pageSize.orient = orient;
    return this;
  }

  docGrid(gridType: DocGridType, linePitch?: number, charSpace?: number) {
    this._docGrid = { gridType, linePitch, charSpace };
    return this;
  }

  header(header: Header) {
    this._header = header;
    return this;
  }

  firstHeader(header: Header) {
    this._firstHeader = header;
    return this;
  }

  evenHeader(header: Header) {
    this._evenHeader = header;
    return this;
  }

  footer(footer: Footer) {
    this._footer = footer;
    return this;
  }

  firstFooter(footer: Footer) {
    this._firstFooter = footer;
    return this;
  }

  evenFooter(footer: Footer) {
    this._evenFooter = footer;
    return this;
  }
}

export type PageOrientationType = "landscape" | "portrait";

export type PageSize = { w: number; h: number; orient?: PageOrientationType };

export type PageMargin = {
  top: number;
  right: number;
  bottom: number;
  left: number;
  header: number;
  footer: number;
  gutter: number;
};

export const defaultPageMargin = () => {
  return {
    top: 1985,
    left: 1701,
    bottom: 1701,
    right: 1701,
    header: 851,
    footer: 992,
    gutter: 0,
  };
};

File: docx-core/src/reader/section_property.rs
use std::io::Read;
use std::str::FromStr;

use xml::attribute::OwnedAttribute;
use xml::reader::{EventReader, XmlEvent};

use super::*;

fn read_page_size(attributes: &[OwnedAttribute]) -> Result<PageSize, ReaderError> {
    let mut size = PageSize::new();
    for a in attributes {
        let local_name = &a.name.local_name;
        match local_name.as_str() {
            "w" => {
                size = size.width(value_to_dax(&a.value)? as u32);
            }
            "h" => {
                size = size.height(value_to_dax(&a.value)? as u32);
            }
            _ => {}
        }
    }
    Ok(size)
}

fn read_page_margin(
    attributes: &[OwnedAttribute],
) -> Result<crate::types::PageMargin, ReaderError> {
    let mut margin = crate::types::PageMargin::new();
    for a in attributes {
        let local_name = &a.name.local_name;
        match local_name.as_str() {
            "top" => {
                margin = margin.top(value_to_dax(&a.value)? as u32);
            }
            "right" => {
                margin = margin.right(value_to_dax(&a.value)? as u32);
            }
            "bottom" => {
                margin = margin.bottom(value_to_dax(&a.value)? as u32);
            }
            "left" => {
                margin = margin.left(value_to_dax(&a.value)? as u32);
            }
            "header" => {
                margin = margin.header(value_to_dax(&a.value)? as u32);
            }
            "footer" => {
                margin = margin.footer(value_to_dax(&a.value)? as u32);
            }
            "gutter" => {
                margin = margin.gutter(value_to_dax(&a.value)? as u32);
            }
            _ => {}
        }
    }
    Ok(margin)
}

fn read_header_or_footer_reference(
    attributes: &[OwnedAttribute],
) -> Result<(String, String), ReaderError> {
    let mut rid = "".to_owned();
    let mut header_type = "default".to_owned();
    for a in attributes {
        let local_name = &a.name.local_name;
        match local_name.as_str() {
            "type" => {
                header_type = a.value.to_owned();
            }
            "id" => {
                rid = a.value.to_owned();
            }
            _ => {}
        }
    }
    Ok((rid, header_type))
}

impl ElementReader for SectionProperty {
    fn read<R: Read>(
        r: &mut EventReader<R>,
        _attrs: &[OwnedAttribute],
    ) -> Result<Self, ReaderError> {
        let mut sp = SectionProperty::new();
        loop {
            let e = r.next();
            match e {
                Ok(XmlEvent::StartElement {
                    attributes, name, ..
                }) => {
                    let e = XMLElement::from_str(&name.local_name).unwrap();
                    match e {
                        XMLElement::PageMargin => {
                            let margin = read_page_margin(&attributes)?;
                            sp = sp.page_margin(margin);
                        }
                        XMLElement::PageSize => {
                            let size = read_page_size(&attributes)?;
                            sp = sp.page_size(size);
                        }
                        XMLElement::DocGrid => {
                            if let Ok(doc_grid) = DocGrid::read(r, &attributes) {
                                sp = sp.doc_grid(doc_grid);
                            }
                        }
                        XMLElement::HeaderReference => {
                            if let Ok((rid, header_type)) =
                                read_header_or_footer_reference(&attributes)
                            {
                                match header_type.as_str() {
                                    "default" => {
                                        sp.header_reference =
                                            Some(HeaderReference::new(header_type, rid));
                                    }
                                    "first" => {
                                        sp.first_header_reference =
                                            Some(HeaderReference::new(header_type, rid));
                                    }
                                    "even" => {
                                        sp.even_header_reference =
                                            Some(HeaderReference::new(header_type, rid));
                                    }
                                    _ => {}
                                }
                            }
                        }
                        XMLElement::FooterReference => {
                            if let Ok((rid, footer_type)) =
                                read_header_or_footer_reference(&attributes)
                            {
                                match footer_type.as_str() {
                                    "default" => {
                                        sp.footer_reference =
                                            Some(FooterReference::new(footer_type, rid));
                                    }
                                    "first" => {
                                        sp.first_footer_reference =
                                            Some(FooterReference::new(footer_type, rid));
                                    }
                                    "even" => {
                                        sp.even_footer_reference =
                                            Some(FooterReference::new(footer_type, rid));
                                    }
                                    _ => {}
                                }
                            }
                        }
                        XMLElement::TitlePg => sp = sp.title_pg(),
                        _ => {}
                    }
                }
                Ok(XmlEvent::EndElement { name, .. }) => {
                    let e = XMLElement::from_str(&name.local_name).unwrap();
                    if e == XMLElement::SectionProperty {
                        return Ok(sp);
                    }
                }
                Err(_) => return Err(ReaderError::XMLReadError),
                _ => {}
            }
        }
    }
}
